File: Interview-Prep/cpp-oop/02-shape-hierarchy/starter.cpp
// Shape hierarchy — fill in the TODOs.
//   g++ -std=c++17 -Wall -Wextra -g starter.cpp -o starter && ./starter
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Shape {
public:
    // TODO: virtual destructor
    // TODO: pure virtual double area() const
    // TODO: pure virtual std::string name() const
};

// TODO: class Circle : public Shape { ... }
// TODO: class Rectangle : public Shape { ... }

// TODO: double totalArea(const std::vector<std::unique_ptr<Shape>>& shapes)

int main() {
    std::vector<std::unique_ptr<Shape>> shapes;
    // TODO: shapes.push_back(std::make_unique<Circle>(2.0)); etc.

    // TODO: print each shape's name + area, then total
    return 0;
}

File: Interview-Prep/python-oop/03-shape-abc/starter.py
"""Shape ABC — fill in the TODOs.  Run: python3 starter.py"""
from abc import ABC, abstractmethod
from dataclasses import dataclass
import math


class Shape(ABC):
    ...  # TODO: @abstractmethod area(self); @abstractmethod name(self)


# TODO: @dataclass class Circle(Shape): radius: float ...
# TODO: @dataclass class Rectangle(Shape): width: float; height: float ...


def total_area(shapes):
    ...  # TODO: sum of areas


if __name__ == "__main__":
    shapes = []  # TODO: [Circle(2.0), Rectangle(3.0, 4.0)]
    for s in shapes:
        print(f"{s.name()} area = {s.area():.2f}")
    print("Total:", total_area(shapes))
